rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,263 @@
+# Composable Pipelines
+
+- **Status**: In-Review
+- **Created**: Aug 19, 2022
+- **ETA**: Sep 19, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Contributors**:
+  - [Anthony Verez](https://github.com/netantho)
+  - [Dominik Lohmann](https://github.com/dominiklohmann)
+  - [Rémi Dettai](https://github.com/rdettai)
+  - [Thomas Peiselt](https://github.com/dispanser)
+- **Discussion**: [PR #2511](https://github.com/tenzir/vast/pull/2511)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      summarize count(dst) group-by src |
+      store parquet /path/to/directory'
+```
+
+The UX would translate seamlessly to other languages, e.g., Python:
+
+```python
+import vast as v
+await v.from("s3://aws")
+       .read_json()
+       .group_by("src") # could also be merged with summarize()
+       .summarize(v.count("dst"))
+       .store_parquet("/path/to/directory")
+```
+
+### Synopsis
+
+The core of this proposal introduces pipeline operators as function with input
+and output types. The following overview introduces the pipeline building
+blocks used throughout this proposal.
+
```

Comment:
Can we add a section here that specifies the possible types as `Void`, `Bytes`, and `Arrow`, and categorizes a pipeline as something that has matching Input and Output types along the pipeline?

E.g., a pipeline with the following operators (in order) has the input type `Void` and the output type `Arrow`: `from<Void, Bytes> | read<Bytes, Arrow> | where<Arrow, Arrow>`.

We can then restrict the individual use sites of pipelines. For example, `vast export FORMAT 'EXPR | Pipeline<Arrow, Arrow>'`, or `vast exec 'Pipeline<Void, Void>'`.

I think this would much help the example section below and help reason about the individual places where we can make use of pipelines.

Reply:
Could you give https://github.com/tenzir/vast/pull/2511/commits/e52eb783878cc95ab21c5dc8e587c9942d1a3cbb read? I tried addressing this concern there.

Reply:
Mostly reads great. I left some individual comments on that, so feel free to resolve this thread.